ocaml/xapi/xapi_network.ml

```diff
@@ -329,3 +329,40 @@ let with_networks_attached_for_vm ~__context ?host ~vm f =
         error "Caught %s while detaching networks" (string_of_exn e)
     end;
     raise e
+
+(* Note that in the revision history is a version of this function
+ * with logic for additional purposes, which may be useful in future. *)
+let assert_can_add_purpose ~__context ~network ~current newval =
+  let sop (*string-of-porpoise*) = Record_util.network_purpose_to_string in
+  let reject str =
+    raise Api_errors.(Server_error (network_incompatible_purposes, [ (sop newval); str ]))
+  in
+  let assert_no_net_has_bad_porpoise bads =
+    (* Sadly we can't use Db.Network.get_refs_where because the expression
+     * type doesn't allow searching for a value inside a list. *)
+    Db.Network.get_all ~__context |>
+    List.iter (fun nwk ->
+      Db.Network.get_purpose ~__context ~self:nwk |>
+      List.iter (fun suspect -> if (List.mem suspect bads) then
+        info "Cannot set new network purpose %s when there is a network with purpose %s" (sop newval) (sop suspect);
+        reject (sop suspect)
+      )
+    )
+  in
+  match newval with
+  | `nbd -> assert_no_net_has_bad_porpoise [`insecure_nbd]
+  | `insecure_nbd -> assert_no_net_has_bad_porpoise [`nbd]
+
+let add_purpose ~__context ~self ~value =
+  let current = Db.Network.get_purpose ~__context ~self in
+  if not (List.mem value current) then (
+    assert_can_add_purpose ~__context ~network:self ~current value;
+    Db.Network.set_purpose ~__context ~self ~value:(value::current)
+  )
```

Comment:
As we discussed we could use the autogenerated`DB.Network.add_to/remove_from_purpose` functions, provided they are idempotent. @robhoes do you know if these throw an error for already added /removed items?

Reply:
They definitely weren't idempotent. I recall that @jonludlam made some changes in this area, but can't remember the details...